README.md
pycard (pronounced picard)
==========================

pycard is a payment card validation library with a simple interface and no
external dependencies.

Installation
------------

```bash
pip install pycard
```

That's it, and there are no dependencies!

Usage
-----

```python
import pycard

# Create a card
card = pycard.Card(
    number='4444333322221111',
    month=1,
    year=2020,
    cvc=123
)

# Validate the card (checks that the card isn't expired and is mod10 valid)
assert card.is_valid

# Perform validation checks individually
assert not card.is_expired
assert card.is_mod10_valid

# The card is a visa
assert card.brand == 'visa'
assert card.friendly_brand == 'Visa'
assert card.mask == 'XXXX-XXXX-XXXX-1111'

# The card is a known test card
assert card.is_test
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-09-03
### Added
- jcb regex and jcb card brand
- 44 new test card number
- pytest, coverage, pylint
- library Exceptions

### Changed
- Convert Card class data dicts to Enum class type
- simplified some part of code
- typo and comments

### Removed
- unicode strings => u'something'
- card repr date
- ExpDate class


## [0.0.1] - 2022-09-02
### Added
- Fork from captain_pycard

tests/test_card.py
"""
All card tests
"""

from pycard.card import Card

card_test_visa = (
    "4000000000000002",
    "4000000000000077",
    "4000000000000259",
    "4000000000001976",
    "4000000000003055",
    "4000000000003220",
    "4000000000005423",
    "4000000000009979",
    "4000000000009987",
    "4000000000009995",
    "4000000360000006",
    "4000000400000008",
    "4000000560000004",
    "4000000760000002",
    "4000001240000000",
    "4000001560000002",
    "4000002080000001",
    "4000002500000003",
    "4000002760000016",
    "4000003440000004",
    "4000003720000005",
    "4000003800000008",
    "4000003920000003",
    "4000004420000006",
    "4000004840000008",
    "4000005280000002",
    "4000005540000008",
    "4000005780000007",
    "4000006200000007",
    "4000006430000009",
    "4000007020000003",
    "4000007240000007",
    "4000007520000008",
    "4000007560000009",
    "4000008260000000",
    "4012888888881881",
    "4444333322221111",
    "4111111111111111",
    "4222222222222",
    "4242424242424242",

)

card_test_mastercard = (
    "30569309025904",
    "3530111333300000",
    "371449635398431",
    "378282246310005",
    "378734493671000",
    "38520000023237",
    "5105105105105100",
    "5105105105105100",
    "5200828282828210",
    "555555555554444",
    "5555555555554444",
    "6011000990139424",
    "6011111111111117",
)

card_test_discover = (
    '6011894492395579',
    '6011388644154687',
    '6011880085013612',
    '6011652795433988',
    '6011375973328347',
)


def card_brand_tester(card_number: str) -> str:
    card = Card(card_number)
    return card.brand()


def test_is_card_brand_visa():
    for elem in card_test_visa:
        assert card_brand_tester(elem) == 'visa'


def test_is_card_brand_mastercard():
    for elem in card_test_visa:
        assert card_brand_tester(elem) == 'mastercard'


def test_is_card_brand_amex():
    for elem in card_test_visa:
        assert card_brand_tester(elem) == 'amex'

pycard/card.py
"""
The class library
"""

from datetime import datetime
from datetime import timedelta

from pycard.data import Brand
from pycard.data import FriendlyBrand
from pycard.data import TESTS_CARDS
from pycard.exceptions import CardNuberNotDigitException
from pycard.exceptions import MaskException


class Card:
    """
    A credit card that may be valid or invalid.
    """

    def __init__(
        self, number: str, month: int = None, year: int = None, cvc: int = None, holder=None
    ):
        """
        Attaches the provided card data and holder to the card after removing
        non-digits from the provided number.
        """
        if not number.isdigit():
            raise CardNuberNotDigitException(
                f"card number {number} contain non digit character(s)"
            )
        self.cvc: int = cvc
        self.holder: int = holder
        self.month: int = month
        self.number: str = number
        self.year: int = year

    def mask(self) -> str:
        """
        Returns the credit card number with each of the number's digits but the
        first six and the last four digits replaced by an X, formatted the way
        they appear on their respective brands' cards.
        """
        # If the card is invalid, return an "invalid" message
        if not self.is_mod10_valid:
            raise MaskException(
                "Unable to generate card mask, due to mod10 invalid"
            )

        # If the card is an Amex, it will have special formatting
        if self.brand() == Brand.amex.name:
            return f"XXXX-XXXXXX-X{self.number[11:15]}"

        # All other cards
        return f"XXXX-XXXX-XXXX-{self.number[12:16]}"

    def brand(self) -> str:
        """
        # Check if the card is of known type and return it
        """
        # Check if the card is of known type
        for brand in Brand:
            if brand.value.match(self.number):
                return brand.name

        return "unknown"

    def friendly_brand(self) -> str:
        """
        Returns the human-friendly brand name of the card.
        """
        for friendly in FriendlyBrand:
            if friendly.name == self.brand():
                return friendly.value
        raise MaskException("Unable to find matching friendly brand")

    def is_test(self) -> bool:
        """
        Returns whether the card's number is a known test number.
        """
        return self.number in TESTS_CARDS

    def is_expired(self) -> bool:
        """
        Returns whether the card is expired.
        """

        today: datetime = datetime.utcnow() - timedelta(hours=11)
        if self.year < today.year or self.month < today.month:
            return True

    def is_valid(self) -> bool:
        """
        Returns whether the card is a valid card for making payments.
        """
        return not self.is_expired and self.is_mod10_valid

    def is_mod10_valid(self) -> bool:
        """
        Returns whether the card's number validates against the mod10
        algorithm (Luhn algorithm), automatically returning False on an empty
        value.
        """

        # Run mod10 on the number
        dub, tot = 0, 0
        for i in range(len(str(self.number)) - 1, -1, -1):
            for c in str((dub + 1) * int(str(self.number)[i])):
                tot += int(c)
            dub = (dub + 1) % 2

        return (tot % 10) == 0


pycard/data.py
"""
All enum and datas used to perform card validation
"""
import re
from enum import Enum


class Brand(Enum):
    """
    All regex associated to card brands
    """

    amex = re.compile(r"^3[47]\d{13}$")
    dankort = re.compile(r"^(5019)\d{12}$")
    diners = re.compile(r"^3(?:0[0-5]|[68][0-9])[0-9]{11}$")
    discover = re.compile(r"^(6011|65\d{2})\d{12}$")
    jcb = re.compile(r"^(?:2131|1800|35\d{3})\d{11}$")
    mastercard = re.compile(
        r"""
                    ^(5[1-5]\d{4}|677189)\d{10}$|  # Traditional 5-series + RU support
                    ^(222[1-9]|2[3-6]\d{2}|27[0-1]\d|2720)\d{12}$  # 2016 2-series
                """,
        re.VERBOSE,
    )
    maestro = re.compile(r"^(?:5[0678]\d\d|6304|6390|67\d\d)\d{8,15}$")
    visa = re.compile(r"^4\d{12}(\d{3})?$")


class FriendlyBrand(Enum):
    """
    More user-friendly card brand matching
    """

    amex = "American Express"
    dankort = "Dankort"
    diners = "Diners Club"
    discover = "Discover"
    jcb = "JCB"
    maestro = "Maestro"
    mastercard = "MasterCard"
    unknown = "unknown"
    visa = "Visa"


TESTS_CARDS = (
    "30569309025904",
    "3530111333300000",
    "371449635398431",
    "378282246310005",
    "378734493671000",
    "38520000023237",
    "4000000000000002",
    "4000000000000077",
    "4000000000000259",
    "4000000000001976",
    "4000000000003055",
    "4000000000003220",
    "4000000000005423",
    "4000000000009979",
    "4000000000009987",
    "4000000000009995",
    "4000000360000006",
    "4000000400000008",
    "4000000560000004",
    "4000000760000002",
    "4000001240000000",
    "4000001560000002",
    "4000002080000001",
    "4000002500000003",
    "4000002760000016",
    "4000003440000004",
    "4000003720000005",
    "4000003800000008",
    "4000003920000003",
    "4000004420000006",
    "4000004840000008",
    "4000005280000002",
    "4000005540000008",
    "4000005780000007",
    "4000006200000007",
    "4000006430000009",
    "4000007020000003",
    "4000007240000007",
    "4000007520000008",
    "4000007560000009",
    "4000008260000000",
    "4012888888881881",
    "4111111111111111",
    "4222222222222",
    "4242424242424242",
    "4444333322221111",
    "5105105105105100",
    "5105105105105100",
    "5200828282828210",
    "555555555554444",
    "5555555555554444",
    "6011000990139424",
    "6011111111111117",
)
